### DISCUSSION TITLE
Amendment XLS: DriftVault — Slippage Rebates for XRP Settlement

### DISCUSSION THREAD

**Discussion by jamestwilkinson** [XRPLF/XRPL-Standards#605] [XLS Ideas (pre standard proposal)]
<html>
<body>
<html><head></head><body><hr>
<h2>title: DriftVault
description: A slippage rebate mechanism for XRP settlement
author: Jimbo
category: Amendment
status: Draft
requires: Escrow, XLS-30 (AMM), XLS-100 (Smart Escrow)
created: 2026-08-22</h2>
<h1>Amendment XLS: "DriftVault" — Slippage Rebates for XRP Settlement on the XRP Ledger</h1>
<h2>Abstract</h2>
<p>DriftVault introduces a native ledger primitive that compensates counterparties for adverse XRP price movement (drift) occurring between the time a payment obligation is quoted/locked and the time it settles. It does this utilising a reserve pool (DriftVault)  that pays a partial, slippage-linked rebate to a claimant when XRP has depreciated against a reference currency over the holding window of a locked payment.</p>

<p> The DriftVault needs to be funded, so a small fee is proposed on transactions benefitting from the feature: 

- a Payment with XRP as the delivered asset
- an EscrowFinish delivering XRP
- an AMM swap where the output is XRP
- a DEX trade where the output is XRP
- an OfferCreate that results in XRP being delivered
- an NFT buy where the buyer pays in XRP
- an IOU redemption where the issuer pays out XRP </p>

<p>The mechanism is designed to lower the effective volatility cost of settling in XRP across any payment with a gap between price commitment and settlement — including but not limited to time-boxed, delegated, or agent-initiated payments — without pegging XRP itself, minting a synthetic stable-asset, or requiring off-ledger intermediaries.</p>
<h2>Motivation</h2>
<p>Any XRP-denominated payment that has a gap between when the price is agreed and when the payment actually settles carries volatility risk that a stablecoin-denominated payment does not. During that gap, XRP's price can move — so the amount committed to no longer matches the amount actually owed by settlement time. A stablecoin payment doesn't have this problem, because its value doesn't move.</p>
<p>This is a general feature of XRP settlement, it applies to routine commerce, business invoicing, and any payment that isn't instantaneously agreed and settled in the same moment — and it may be especially pronounced for AI agents making payments under a budget or spending limit set in advance (e.g. "book this trip, but keep it under $2,000, and you have a few hours to decide"), where the commitment-to-settlement window is explicit and often longer.</p>
<p>That makes XRP a less attractive choice across this whole range of payments, and speeding up settlement doesn't fix it — even a fast settlement still leaves a window where the price can drift before the payment closes out.</p>
<p>Today, participants who want price-stable settlement default to stablecoins (e.g. RLUSD), subordinating XRP primarily to a bridge/liquidity role rather than as a direct settlement medium for commerce generally. This runs against the principle that a blockchain's native asset should be the default choice for mediating core transactions on-chain, rather than routing users toward secondary assets like stablecoins.</p>
<p>The motivation for this amendment is to reinforce native settlement as the preferred mechanism ahead of synthetic settlement. XRP settlement carries properties a stablecoin does not:</p>
<ul>
<li>deep, protocol-native liquidity, not dependent on an issuer's market-making</li>
<li>deterministic transaction rules</li>
<li>protocol-level fee mechanics</li>
<li>no freeze capability</li>
<li>no issuer default risk</li>
<li>universal acceptance across all XRPL nodes, with no trust-line requirement</li>
<li>a counterparty-free asset</li>
</ul>
<p>This proposal aims to restore XRP's original design intent as the preeminent settlement asset on the chain — without forcing payers to use XRP, and while making native settlement a more competitive choice against the risks and frictions of synthetic settlement.</p>
<p>DriftVault aims to narrow the gap by utilising smart escrow functionality to make XRP settlement risk partially self-insuring at the protocol level, funded by the ecosystem's own transaction volume rather than by a third-party underwriter.</p>
<p>This is deliberately <strong>not</strong> a proposal to stabilize XRP's price, peg it, or create a synthetic dollar token. It is a compensation mechanism scoped to a bounded holding window, funded by a transparent, auditable on-ledger reserve sufficient to redress the balance and protect this aspect of XRP's utility.</p>
<h2>Specification</h2>
<h3>Overview of flow</h3>
<ol>
<li>Payer decides to send a transaction.</li>
<li>DriftVault checks the price oracle.</li>
<li>DriftVault calculates the required XRP, plus a buffer, plus the DV fee.</li>
<li>Payer locks XRP into a <code>DVLock</code> — a smart escrow object that covers the future payment, recording the reference price (<code>ReferencePrice</code>) at lock time (<code>t0</code>).</li>
<li>The future payment event occurs (e.g. a time condition or a mandate-defined trigger is met).</li>
<li>DriftVault re-checks the price oracle at settlement time (<code>t1</code>).</li>
<li>DriftVault converts the locked XRP to the target currency value, with settlement still executed in XRP.</li>
<li>The recipient receives the XRP settlement, including any slippage-linked rebate owed from the <code>DriftVault</code> pool.</li>
<li>Any excess XRP is returned to the payer.</li>
<li>DriftVault retains the DV fee for pool sustainability.</li>
</ol>
<p>Settlement (steps 5–10) is triggered automatically once the payment condition is met, rather than requiring the destination account to actively submit a claim. This "push" model — where any account (payer, recipient, or a relayer/keeper) can submit the settlement transaction once the condition is satisfied — mirrors how <code>EscrowFinish</code> can already be called by any account, not only the destination, once a time or crypto-condition is met. This suits unattended, agent-initiated payments, where no human is necessarily present to trigger a manual claim.</p>
<h3>New ledger entry types</h3>
<p><strong><code>DriftVault</code></strong>
Represents a compensation reserve pool. Fields:</p>

Field | Type | Description
-- | -- | --
Owner | AccountID | Account (protocol/treasury) controlling pool parameters
Balance | Amount | Current XRP or token balance available for payout
CompensationRate | UInt16 (basis points) | Current payout rate applied to realized losses, e.g. 7000 = 70%
TrailingPayout30d | Amount | Rolling 30-day payout total, used to compute solvency ratio
ReserveRatioFloor | UInt16 | Minimum acceptable (Balance ÷ TrailingPayout30d) before CompensationRate begins decaying
MaxClaimPerTx | Amount | Per-transaction payout cap
MaxClaimPerAccount | Amount | Rolling-window payout cap per claimant account

<h3>New transaction types</h3>
<p><strong><code>DriftVaultFund</code></strong>
Deposits the DV fee into a <code>DriftVault</code> as part of transaction processing. Rather than being a standalone user-submitted transaction, this is a protocol-level side effect: the DV fee is assessed on qualifying transactions (per the fee model in the Abstract) and routed into the associated <code>DriftVault.Balance</code> at the same time the transaction is processed.</p>
<p><strong><code>DVLockCreate</code></strong>
Creates a <code>DVLock</code>, analogous to <code>EscrowCreate</code>. Requires <code>Amount</code>, <code>Destination</code>, <code>VaultID</code>, <code>Expiration</code>, and reads <code>ReferencePrice</code> from the specified <code>Oracle</code> at execution time.</p>
<p><strong><code>DVSettle</code></strong>
Analogous to <code>EscrowFinish</code>, but callable by any account (payer, recipient, or a relayer/keeper) once the payment condition is met — not destination-only. On execution:</p>
<ul>
<li>Reads current price from <code>OracleID</code> using a time-weighted average (TWAP) over a short window (e.g. last 3–5 ledger closes) rather than a single spot read, to reduce timing-manipulation risk.</li>
<li>Computes <code>loss = max(0, (ReferencePrice - CurrentPrice) × Amount)</code>.</li>
<li>Computes <code>payout = min(loss × CompensationRate, MaxClaimPerTx, remaining MaxClaimPerAccount headroom)</code>.</li>
<li>Transfers <code>Amount</code> (the underlying locked XRP) to <code>Destination</code>.</li>
<li>Transfers <code>payout</code> from <code>DriftVault.Balance</code> to <code>Destination</code>, if <code>Balance</code> is sufficient; otherwise pays out the available remainder and the settlement partially fails-soft (no full transaction failure, since the underlying payment must still complete).</li>
<li>Returns any excess locked XRP to the payer.</li>
</ul>
<p><strong><code>DVLockCancel</code></strong>
Analogous to <code>EscrowCancel</code>. Returns locked XRP to the payer if <code>Expiration</code> has passed and no settlement was submitted.</p>
<p><strong><code>DriftVaultVote</code></strong> <em>(optional, governance extension)</em>
Allows accounts holding a stake in the vault (if implemented as a share-based pool similar to <code>Vault</code>/<code>LoanBroker</code> LP mechanics) to vote on <code>CompensationRate</code> and <code>ReserveRatioFloor</code> parameters, paralleling <code>AMMVote</code>.</p>
<h3>CompensationRate decay</h3>
<p><code>CompensationRate</code> is not static. It is recalculated periodically (e.g. once per N ledgers) as a function of the pool's solvency ratio:</p>
<pre><code>solvency_ratio = DriftVault.Balance / DriftVault.TrailingPayout30d
</code></pre>
<p>If <code>solvency_ratio</code> falls below <code>ReserveRatioFloor</code>, <code>CompensationRate</code> decays smoothly toward a floor (e.g. 10%) rather than dropping abruptly to zero, using a bonding-curve-style function. This avoids a hard cliff in payout expectations during sustained XRP downtrends.</p>
<h2>Rationale</h2>
<ul>
<li><strong>Loss-conditional, not flat, rebate.</strong> Compensation is only paid on realized adverse drift, not on every settlement. This is materially more capital-efficient than a flat per-transaction rebate and reduces exposure to wash-trading/farming relative to unconditional payouts.</li>
<li><strong>Partial (&lt;100%) compensation rate.</strong> Leaving some residual risk with the claimant preserves incentive alignment and avoids the vault underwriting unlimited one-sided risk, which would be undercapitalized in any sustained downtrend.</li>
<li><strong>TWAP over spot oracle reads.</strong> A single spot price at settlement is manipulable by whichever account chooses submission timing under the push model; a short TWAP window meaningfully raises the cost of gaming the payout.</li>
<li><strong>Smart-escrow-derived design.</strong> Building on <code>Escrow</code> semantics, extended via XLS-100 Smart Escrow functionality, minimizes protocol surface area and reuses ledger mechanics (reserves, expiration, conditional execution) the network already has security experience with, while enabling the variable, oracle-dependent payout calculation that plain <code>Escrow</code> cannot perform on its own.</li>
<li><strong>Why not just use a stablecoin (e.g. RLUSD)?</strong> DriftVault does not claim to be more capital-efficient than direct stablecoin settlement for pure price-stability purposes. It is a deliberate, transparent, protocol-level mechanism to reinforce XRP as the network's native settlement asset across the general run of on-chain commerce — not only time-boxed or agent-mediated payments — funded by the ecosystem's own transaction volume, not a claim that this design is more capital-efficient than native price stability.</li>
</ul>
<h2>Backwards Compatibility</h2>
<p>DriftVault introduces new ledger entry types and transaction types; it does not modify existing <code>Escrow</code>, <code>AMM</code>, or <code>Oracle</code> behaviour. Existing accounts, escrows, and transaction flows are unaffected. Adoption is opt-in: a payer/destination pair must explicitly choose to route a payment through a <code>DVLock</code> rather than a standard <code>Payment</code> or <code>Escrow</code>. The new DV fee, however, is proposed to apply to only qualifying transactions (per the Abstract's fee-scope discussion).</p>
<h2>Security Considerations</h2>
<ul>
<li><strong>Oracle manipulation.</strong> TWAP mitigates but does not eliminate oracle-timing risk, particularly for low-liquidity <code>Oracle</code> price feeds. Feeds used for <code>DVLock</code> should meet a minimum liquidity/update-frequency bar (to be defined). Under the push settlement model, this risk extends to any account eligible to submit <code>DVSettle</code>, not only the destination, since any submitter can choose a favorable TWAP window to trigger within.</li>
<li><strong>Pool insolvency in sustained downtrends.</strong> The <code>CompensationRate</code> decay function bounds this but does not eliminate it; a treasury backstop or fixed-supply cap on total outstanding <code>MaxClaimPerAccount</code> exposure across the ecosystem may be needed as an additional safeguard.</li>
<li><strong>Sybil/wash exposure.</strong> Per-account and per-transaction caps reduce, but a full mitigation likely requires binding claims to <code>Credential</code>-verified accounts (per XLS-70) or <code>PermissionedDomain</code> (XLS-80) participation, especially at higher <code>CompensationRate</code> values.</li>
<li><strong>Fee scope and network-wide impact.</strong>The DV fee applies only to qualifying XRP‑settled transactions, as defined in the Abstract (e.g., Payments delivering XRP, EscrowFinish delivering XRP, AMM swaps where the output is XRP). This ensures that only transactions benefiting from DriftVault contribute to its sustainability. Because the fee is scoped and opt‑in at the settlement level, it does not constitute a network‑wide base‑fee change, and does not impose costs on accounts or applications that do not use DVLock or XRP‑settled flows. The fee‑routing logic and DriftVault.Balance must remain fully auditable on‑ledger to maintain trust in the mechanism’s solvency.</li>
</ul>
<h2>Open Questions</h2>
<ul>
<li>Should the DV fee apply to all XRP-denominated transactions in the suggested list, or otherwise? </li>
<li>Should DriftVault pools be per-application/per-platform, or a single ecosystem-wide shared pool? A shared pool has deeper reserves but weaker incentive-alignment between the fee source and the payout recipient.</li>
<li>Should DriftVaultVote (LP-style governance) be in scope for the initial amendment, or deferred to a follow-up XLS once the core lock/settle mechanism has been live and observed?</li>
<li>Should MaxClaimPerAccount be enforced natively at the protocol level, or left to application-layer enforcement on top of the primitive?</li>
</ul></body></html>
</body>
</html>
